include/nccl_ofi_idpool.h

```diff
@@ -5,90 +5,78 @@
-	   that the ID corresponding to its index is available.*/
-	uint64_t *ids;
 
-	/* Lock for concurrency */
-	pthread_mutex_t lock;
-} nccl_ofi_idpool_t;
+	/* Disable implicit copy constructor and asignment operator */
+	nccl_ofi_idpool_t(const nccl_ofi_idpool_t&) = delete;
+	nccl_ofi_idpool_t& operator=(const nccl_ofi_idpool_t&) = delete;
 
-/*
- * @brief	Initialize pool of IDs
- *
- * Allocates and initializes a nccl_ofi_idpool_t object, marking all
- * IDs as available.
- *
- * @param	idpool_p
- *		Return value with the ID pool pointer allocated
- * @param	size
- *		Size of the id pool (number of IDs)
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_init(nccl_ofi_idpool_t *idpool, size_t size);
 
-/*
- * @brief	Allocate an ID
- *
- * Extract an available ID from the ID pool, mark the ID as
- * unavailable in the pool, and return extracted ID. No-op in case
- * no ID was available.
- *
- * This operation is locked by the ID pool's internal lock.
- *
- * @param	idpool
- *		The ID pool
- * @return	the extracted ID (zero-based) on success,
- *		negative value on error
- */
-int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool);
+	/*
+	 * @brief	Allocate an ID
+	 *
+	 * Extract an available ID from the ID pool, mark the ID as
+	 * unavailable in the pool, and return extracted ID. Return error
+	 * code in case no ID was available.
+	 *
+	 * This operation is locked by the ID pool's internal lock.
+	 *
+	 * @return	the extracted ID (zero-based) on success,
+	 *		negative value on error
+	 */
+	ssize_t allocate_id();
 
-/*
- * @brief	Free an ID from the pool
- *
- * Return input ID into the pool.
- *
- * This operation is locked by the ID pool's internal lock.
- *
- * @param	idpool
- *		The ID pool
- * @param	id
- *		The ID to release (zero-based)
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_free_id(nccl_ofi_idpool_t *idpool, size_t id);
 
-/*
- * @brief	Release pool of IDs and free resources
- *
- * Releases a nccl_ofi_idpool_t object and frees allocated memory.
- *
- * @param	idpool_p
- *		Pointer to the ID pool, it will be set to NULL on success
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_fini(nccl_ofi_idpool_t *idpool);
+	/*
+	 * @brief	Free an ID from the pool
+	 *
+	 * Return input ID into the pool.
+	 *
+	 * This operation is locked by the ID pool's internal lock.
+	 *
+	 * @param	id
+	 *		The ID to release (zero-based)
+	 * @return	0 on success
+	 *		non-zero on error
+	 */
+	int free_id(size_t id);
 
 
-/*
- * @brief        Check if an idpool has been initialized with a size
- *               other than 0.
- */
-static inline bool nccl_ofi_idpool_active(nccl_ofi_idpool_t *idpool) {
-	return (idpool->size != 0);
-}
+	/* Return number of IDs in the id pool */
+	size_t get_size();
+
+
+	/* Return the idpool element value of a valid vector index, return 
+	   negative error code on out-of-bound index */
+	uint64_t get_element(size_t index);
```

Comment:
I'm not a huge fan of making public interface functions just for testing.  I think instead I'd make the variables protected instead of private, and write a class in your test function that inherits from idpool and provides accessors to idpool.  That way, you don't have this weird interface that confuses everyone.